### PR TITLE
[Storage] Support for storage info in cluster table's metadata

### LINKED ISSUE #1203: [Storage] VM loses S3 mount after stop/start
To reproduce: make `cluster.yml`:

```
resources:
  cloud: aws
  instance_type: g4dn.xlarge
  disk_size: 500

workdir: .

num_nodes: 4

file_mounts:
  /dataset:
    source: s3://raysort-kl
    store: s3

setup: |
  echo "Hello, SkyPilot!"

run: |
  echo "Hello, SkyPilot!"
```

Then

```
sky launch cluster.yml -c g4
# login, verify content of /dataset is present
sky stop -y g4
sky start -y g4
# login, /dataset becomes empty
```

I _think_ this also happens on a single-node VM but I haven't verified it yet.

### PR DESCRIPTION
**Pull request by landscapepainter**
This PR resolves #1203.

Currently, when we run `sky stop <cluster-name>` on a cluster that has cloud storage mounted, the mounting will be lost when restarted with `sky start <cluster-name>` as the mounting command does not get ran. 

In this PR, I save the `storage_mounts` dictionary as a metadata in `cluster` table's `storage_mounts_metadata` when `sky launch` is initially ran. Then, this is loaded to be used to re-mount when `sky start <cluster-name>` is ran.

Tested (run the relevant ones):

- [x] Code formatting: `bash format.sh`
- [x] Manual tests on `gcsfuse`, `goofys`, and `rclone`.
- [x] Relevant individual smoke tests: `pytest tests/test_smoke.py::TestStorageWithCredentials` 
- [x] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_file_mounts`
- [x] `state.db` created with `sky launch` in `master` branch, `sky stop`/`start` ran from PR branch
- [x] `state.db`created with `sky launch` in `master` branch, `sky stop` in master branch, `sky start` in PR branch
- [x] Running `sky storage delete` where storage has 3 stores and 1 of those are externally deleted
- [x] Running `sky start` with a cluster that has a bucket mounted and deleted externally
- [x] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_aws_storage_mounts_with_stop --aws`
- [x] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_gcp_storage_mounts_with_stop --gcp`
- [x] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_kubernetes_storage_mounts_with_stop --kubernetes`

### CODE REVIEW

**Review by landscapepainter — COMMENTED** [review]
This is now ready for a review.

File: sky/core.py [hunk 1]
```diff
@@ -208,6 +208,8 @@ def _start(
                                stream_logs=True,
                                cluster_name=cluster_name,
                                retry_until_up=retry_until_up)
+    storage_mounts = backend.get_cluster_metadata_storage_mounts(handle)
+    backend.sync_file_mounts(handle, None, storage_mounts)
```
Comment (romilbhardwaj) [inline sky/core.py:212]: This will also trigger our file mount syncing code. This can be problematic - e.g., if a user's local path has been deleted, this will likely raise an error and prevent the user from starting a stopped cluster. 

Since rsync file_mounts and `COPY` mode objects would anyway be persistent on the remote disk volume, is there a way we can execute only `MOUNT` mode file_mounts and skip the local->remote upload for them, if any?

**Review by romilbhardwaj — COMMENTED** [review]
Thanks @landscapepainter! Took a high level look at the code. Haven't tested yet.

File: sky/core.py (hunk 1 above)
Comment (landscapepainter) [reply sky/core.py:212]: It seems we're specifically executing in `MOUNT` mode only here, as we pass `None` as an argument to `backend.sync_file_mounts()`. When `file_mounts` parameter is set to `None`, `_execute_file_mounts()` will return immediately without taking any actions. Please correct me if I'm overlooking anything with this appraoch!

File: sky/backends/cloud_vm_ray_backend.py [hunk 2]
```diff
@@ -4240,6 +4241,45 @@ def _execute_storage_mounts(self, handle: CloudVmRayResourceHandle,
...
+        cluster_name = handle.cluster_name
+        cluster_metadata = global_user_state.get_cluster_metadata(cluster_name)
+        if cluster_metadata is None:
+            return None
+        if not 'storage' in cluster_metadata.keys():
+            cluster_metadata['storage'] = {}
+        for _, storage_obj in storage_mounts.items():
+            for _, store_obj in storage_obj.stores.items():
+                # Update some of the non-picklabe attributes of store instances
+                store_obj.make_picklable()
```
[the remark this answers is not in the document]
Comment (landscapepainter) [reply sky/backends/cloud_vm_ray_backend.py:4260]: Thanks for pointing this out!

File: sky/backends/cloud_vm_ray_backend.py [hunk 3]
```diff
@@ -4240,6 +4241,45 @@ def _execute_storage_mounts(self, handle: CloudVmRayResourceHandle,
...
 
+    def _set_cluster_metadata_storage_mounts(
+            self, handle: CloudVmRayResourceHandle,
+            storage_mounts: Dict[Path, storage_lib.Storage]) -> None:
+        """Sets 'storage_mounts' object in cluster's storage metadata"""
+        if not storage_mounts:
+            return
+
+        cluster_name = handle.cluster_name
+        cluster_metadata = global_user_state.get_cluster_metadata(cluster_name)
```
[the remark this answers is not in the document]
Comment (landscapepainter) [reply sky/backends/cloud_vm_ray_backend.py:4252]: To add on, the current implementation I have, it stores `storage_mounts` dictionary, which is `dst[str] : Storage` obj. pair in `cluster`'s `metadata`, which is useful as I can just pass it into `sync_file_mounts` api just like we run it on `sky launch`. We need to consider a way to store dst and keep it as a pair with `Storage` obj. if we want to store only `Storage` object in handle.

**Comment by landscapepainter** [comment]
@romilbhardwaj This is ready for a look. Did some additional backwards compatibility tests as well.

File: sky/data/storage.py [hunk 4]
```diff
@@ -713,6 +693,57 @@ def validate_name(name):
             f'Validation failed for storage source {self.source}, name '
             f'{self.name} and mode {self.mode}. Please check the arguments.')
 
+    def _add_store_from_metadata(self, sky_stores) -> None:
```
Comment (romilbhardwaj) [inline sky/data/storage.py:696]: Nice to add type hints for `sky_stores`

File: sky/core.py [hunk 5]
```diff
@@ -208,6 +208,8 @@ def _start(
                                stream_logs=True,
                                cluster_name=cluster_name,
                                retry_until_up=retry_until_up)
+    storage_mounts = backend.get_storage_mounts_metadata(handle.cluster_name)
+    backend.sync_file_mounts(handle, None, storage_mounts)
```
Comment (romilbhardwaj) [inline sky/core.py:212]: There seems to be a bug if I try to `sky start` with a bucket that has been deleted. We may need to add some handling for checking if the storage object/bucket still exists and throw and a cleaner error if not:

1. `sky launch -c clus` this YAML:
```
file_mounts:
  /bucket:
    name: romil-test-bucket
    store: gcs
    mode: MOUNT
```
2. `sky stop clus`
3. `sky storage delete romil-test-bucket`
4. `sky start clus` fails with:

```
I 10-02 12:14:03 cloud_vm_ray_backend.py:4550] Processing 1 storage mount.
Traceback (most recent call last):
  File "/Users/romilb/tools/anaconda3/bin/sky", line 8, in <module>
    sys.exit(cli())
  File "/Users/romilb/tools/anaconda3/lib/python3.9/site-packages/click/core.py", line 1128, in __call__
    return self.main(*args, **kwargs)
  File "/Users/romilb/tools/anaconda3/lib/python3.9/site-packages/click/core.py", line 1053, in main
    rv = self.invoke(ctx)
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/utils/common_utils.py", line 311, in _record
    return f(*args, **kwargs)
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/cli.py", line 1162, in invoke
    return super().invoke(ctx)
  File "/Users/romilb/tools/anaconda3/lib/python3.9/site-packages/click/core.py", line 1659, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/romilb/tools/anaconda3/lib/python3.9/site-packages/click/core.py", line 1395, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/romilb/tools/anaconda3/lib/python3.9/site-packages/click/core.py", line 754, in invoke
    return __callback(*args, **kwargs)
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/utils/common_utils.py", line 332, in _record
    return f(*args, **kwargs)
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/cli.py", line 2563, in start
    core.start(name,
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/utils/common_utils.py", line 332, in _record
    return f(*args, **kwargs)
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/core.py", line 274, in start
    _start(cluster_name,
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/core.py", line 212, in _start
    backend.sync_file_mounts(handle, None, storage_mounts)
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/utils/common_utils.py", line 332, in _record
    return f(*args, **kwargs)
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/utils/common_utils.py", line 311, in _record
    return f(*args, **kwargs)
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/backends/backend.py", line 72, in sync_file_mounts
    return self._sync_file_mounts(handle, all_file_mounts, storage_mounts)
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/backends/cloud_vm_ray_backend.py", line 3131, in _sync_file_mounts
    self._execute_storage_mounts(handle, storage_mounts)
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/backends/cloud_vm_ray_backend.py", line 4567, in _execute_storage_mounts
    mount_cmd = store.mount_command(dst)
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/data/storage.py", line 1708, in mount_command
    f'{self.bucket.name} {mount_path}')
AttributeError: 'NoneType' object has no attribute 'name'
```

**Comment by romilbhardwaj** [comment]
Additionally, can you also run `test_smoke.py::test_file_mounts`?

**Comment by landscapepainter** [comment]
@romilbhardwaj Additional testing done for `sky storage delete` and `sky start` after implementing error raise from `_get_bucket` when the bucket is externally deleted while `sync_on_reconstruction` is set to `False`.

- [x] Running sky storage delete where storage has 3 stores and 1 of those are externally deleted
- [ ] Running sky start with a cluster that has a bucket mounted and deleted externally


Following is a result from running `sky start` after externally deleting the mounting bucket.
```
$ sky start doyoung-test -y
I 10-10 05:34:43 cloud_vm_ray_backend.py:1449] To view detailed progress: tail -n100 -f /home/gcpuser/sky_logs/sky-2023-10-10-05-34-43-492829/provision.log
I 10-10 05:34:43 cloud_vm_ray_backend.py:1277] Cluster 'doyoung-test' (status: STOPPED) was previously launched in GCP us-central1. Relaunching in that region.
I 10-10 05:34:47 cloud_vm_ray_backend.py:1887] Launching on GCP us-central1 (us-central1-a)
I 10-10 05:35:21 log_utils.py:45] Head node is up.
I 10-10 05:35:58 cloud_vm_ray_backend.py:1692] Successfully provisioned or found existing VM.
I 10-10 05:36:07 cloud_vm_ray_backend.py:4587] Processing 1 storage mount.
sky.exceptions.StorageExternalDeletionError: The bucket, 'gcs-mount-testing', could not be mounted on cluster 'doyoung-test'. Please verify that the bucket exists.
```

File: sky/data/storage.py [hunk 6]
```diff
@@ -713,6 +691,63 @@ def validate_name(name):
...
+                        s_metadata,
+                        source=self.source,
+                        sync_on_reconstruction=self.sync_on_reconstruction)
+                else:
+                    with ux_utils.print_exception_no_traceback():
+                        raise ValueError(f'Unknown store type: {s_type}')
+            # Following error is raised from _get_bucket and caught only when
+            # an externally removed storage is attempted to be fetched.
+            except exceptions.StorageExternalDeletionError:
+                continue
```
Comment (romilbhardwaj) [inline sky/data/storage.py:732]: Would be nice to add a quick `logger.debug(...)` to document this exception before we `continue`.

File: tests/test_smoke.py [hunk 7]
```diff
@@ -3225,7 +3225,7 @@ def test_nonexistent_bucket(self, nonexist_bucket_url):
 
         with pytest.raises(
                 sky.exceptions.StorageBucketGetError,
-                match='Attempted to connect to a non-existent bucket'):
```
Comment (romilbhardwaj) [inline tests/test_smoke.py:3228]: is it possible to add a smoke test like `test_kubernetes/aws/gcp_storage_mounts`, except changed to `test_storage_mounts_with_stop` to test the stop start functionality? Basically:
1. Launch `tests/test_yamls/test_storage_mounting.yaml.j2` on `generic_cloud`
2. Stop the cluster
3. Start again
4. See if mounting works succesfully by both reading and writing files to `MOUNT` paths.

**Review by romilbhardwaj — COMMENTED** [review]
Thanks @landscapepainter! Left some minor comments, should be good to go after they are fixed. Please also run file mounting related smoke tests and some manual backward compatibility tests to make sure the updates in `global_user_state` don't break for existing storages/clusters.

File: sky/data/storage.py [hunk 8]
```diff
@@ -2007,8 +2022,10 @@ def _get_bucket(self) -> Tuple[StorageHandle, bool]:
           3) Create and return a new bucket otherwise
 
         Raises:
-            StorageBucketCreateError: If creating the bucket fails
```
Comment (romilbhardwaj) [inline sky/data/storage.py:1964]: Same as for S3 above

File: sky/data/storage.py [hunk 9]
```diff
@@ -1215,8 +1226,10 @@ def _get_bucket(self) -> Tuple[Optional[StorageHandle], bool]:
...
```
Comment (romilbhardwaj) [inline sky/data/storage.py:1183]: Why is this removed? I think this can still be raised in `_create_s3_bucket`?

File: sky/data/storage.py [hunk 10]
```diff
@@ -1657,8 +1670,10 @@ def _get_bucket(self) -> Tuple[StorageHandle, bool]:
...
```
Comment (romilbhardwaj) [inline sky/data/storage.py:1619]: Same as for S3 above

**Review by romilbhardwaj — APPROVED** [review]
Thanks @landscapepainter! Left one comment - should be good to go if all smoke tests and manual backward compatibility tests pass.

**Review by romilbhardwaj — CHANGES_REQUESTED** [review]
Ran into an issue with backward compatibility:

1. On master branch, `sky launch -c test task.yaml --cloud gcp` this YAML:
```
file_mounts:
  /nfs:
    name: romil-test-bucket
    store: s3
    mode: MOUNT

run: |
  ls /nfs
```
2. Switch to this branch, run `sky launch -c test task.yaml --cloud gcp` again. (this step may not be required, but I ran it)
3. Stop cluster with `sky stop test`
4. `sky start test` fails with:
```
I 11-07 17:41:15 cloud_vm_ray_backend.py:1692] Successfully provisioned or found existing VM.
D 11-07 17:41:15 cloud_vm_ray_backend.py:3001] Checking if skylet is running on the head node.
Traceback (most recent call last):
  File "/Users/romilb/tools/anaconda3/bin/sky", line 8, in <module>
    sys.exit(cli())
  File "/Users/romilb/tools/anaconda3/lib/python3.9/site-packages/click/core.py", line 1128, in __call__
    return self.main(*args, **kwargs)
  File "/Users/romilb/tools/anaconda3/lib/python3.9/site-packages/click/core.py", line 1053, in main
    rv = self.invoke(ctx)
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/utils/common_utils.py", line 311, in _record
    return f(*args, **kwargs)
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/cli.py", line 1162, in invoke
    return super().invoke(ctx)
  File "/Users/romilb/tools/anaconda3/lib/python3.9/site-packages/click/core.py", line 1659, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/romilb/tools/anaconda3/lib/python3.9/site-packages/click/core.py", line 1395, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/romilb/tools/anaconda3/lib/python3.9/site-packages/click/core.py", line 754, in invoke
    return __callback(*args, **kwargs)
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/utils/common_utils.py", line 332, in _record
    return f(*args, **kwargs)
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/cli.py", line 2573, in start
    core.start(name,
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/utils/common_utils.py", line 332, in _record
    return f(*args, **kwargs)
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/core.py", line 278, in start
    _start(cluster_name,
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/core.py", line 211, in _start
    storage_mounts = backend.get_storage_mounts_metadata(handle.cluster_name)
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/backends/cloud_vm_ray_backend.py", line 4708, in get_storage_mounts_metadata
    storage_mounts[dst] = storage_lib.Storage.from_metadata(
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/data/storage.py", line 760, in from_metadata
    mode=override_args.get('mode', metadata.mode))
AttributeError: 'StorageMetadata' object has no attribute 'mode'
```

**Comment by landscapepainter** [comment]
@romilbhardwaj Re-tested on all the following cases:

- [x] Relevant individual smoke tests: `pytest tests/test_smoke.py::TestStorageWithCredentials` 
- [x] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_file_mounts`

Below is the backwards compatibility test you commented about. Confirmed that it passes now. Thanks for the catch.
- [x] `state.db` created with `sky launch` in `master` branch, `sky stop`/`start` ran from PR branch
- [x] `state.db`created with `sky launch` in `master` branch, `sky stop` in master branch, `sky start` in PR branch

edge cases
- [x] Running `sky storage delete` where storage has 3 stores and 1 of those are externally deleted
- [x] Running `sky start` with a cluster that has a bucket mounted and deleted externally

Newly added smoke tests:
- [x] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_aws_storage_mounts_with_stop --aws`
- [x] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_gcp_storage_mounts_with_stop --gcp`
- [x] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_kubernetes_storage_mounts_with_stop --kubernetes`

**Review by romilbhardwaj — APPROVED** [review]
Thanks for the extensive tests @landscapepainter! Should be good to go now.
